source/host_resolver.c

```diff
@@ -325,11 +366,78 @@ static void s_clear_default_resolver_entry_table(struct default_host_resolver *r
     aws_hash_table_clear(table);
 }
 
-/* TODO: add a callback to notify when purge has completed */
 static int resolver_purge_cache(struct aws_host_resolver *resolver) {
     struct default_host_resolver *default_host_resolver = resolver->impl;
     aws_mutex_lock(&default_host_resolver->resolver_lock);
-    s_clear_default_resolver_entry_table(default_host_resolver);
+    s_clear_default_resolver_entry_table_synced(default_host_resolver);
+    aws_mutex_unlock(&default_host_resolver->resolver_lock);
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_purge_host_cache_callback_task(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)status;
+    struct host_purge_callback_options *options = arg;
+    aws_mem_release(options->allocator, task);
+    aws_ref_count_release(&options->ref_count);
+}
+
+static void s_sechdule_purge_cache_callback_async(
+    struct default_host_resolver *default_host_resolver,
+    aws_simple_completion_callback *on_purge_cache_complete_callback,
+    void *user_data) {
+
+    struct host_purge_callback_options *purge_callback_options = s_host_purge_callback_options_new(
+        default_host_resolver->allocator, on_purge_cache_complete_callback, user_data);
+    purge_callback_options->allocator = default_host_resolver->allocator;
+
+    struct aws_task *task = aws_mem_calloc(default_host_resolver->allocator, 1, sizeof(struct aws_task));
+    aws_task_init(task, s_purge_host_cache_callback_task, purge_callback_options, "async_purge_host_address_task");
+
+    struct aws_event_loop *loop = aws_event_loop_group_get_next_loop(default_host_resolver->event_loop_group);
+    AWS_FATAL_ASSERT(loop != NULL);
+    aws_event_loop_schedule_task_now(loop, task);
+}
+
+static int resolver_purge_cache_with_callback(
+    struct aws_host_resolver *resolver,
+    aws_simple_completion_callback *on_purge_cache_complete_callback,
+    void *user_data) {
+
+    if (!on_purge_cache_complete_callback) {
+        return resolver_purge_cache(resolver);
+    }
+
+    struct default_host_resolver *default_host_resolver = resolver->impl;
+    aws_mutex_lock(&default_host_resolver->resolver_lock);
+    struct aws_hash_table *table = &default_host_resolver->host_entry_table;
+
+    /* Success if hash table is already empty. */
+    if (!aws_hash_table_get_entry_count(table)) {
```

Comment:
super trivial: for clarity
```suggestion
    if (aws_hash_table_get_entry_count(table) == 0) {
```